File: src/Behavioral_Patterns/Command/command.py
# Here is an example of the Command design pattern in Python:


# Command interface
class Command:
    def execute(self):
        pass

# Concrete command classes
class LightOnCommand(Command):
    def __init__(self, light):
        self.light = light
    
    def execute(self):
        self.light.switch_on()

class LightOffCommand(Command):
    def __init__(self, light):
        self.light = light
    
    def execute(self):
        self.light.switch_off()

# Receiver class
class Light:
    def switch_on(self):
        print("Light is switched on.")
    
    def switch_off(self):
        print("Light is switched off.")

# Invoker class
class RemoteControl:
    def __init__(self):
        self.commands = {}
    
    def set_command(self, button, command):
        self.commands[button] = command
    
    def press_button(self, button):
        if button in self.commands:
            self.commands[button].execute()

# Client code
light = Light()
light_on_command = LightOnCommand(light)
light_off_command = LightOffCommand(light)

remote_control = RemoteControl()
remote_control.set_command("ON", light_on_command)
remote_control.set_command("OFF", light_off_command)

remote_control.press_button("ON")
remote_control.press_button("OFF")


# In this example, the Command design pattern is used to control a light using a remote control. The `Command` interface defines the `execute` method which is implemented by concrete command classes (`LightOnCommand` and `LightOffCommand`). These commands encapsulate the receiver object (`Light`) and its methods (`switch_on` and `switch_off`). The `RemoteControl` class acts as an invoker which stores the commands and can execute them when a button is pressed. 

# This design pattern is useful in applications where we want to decouple the sender and receiver of a request, and enable the sender to choose which command to execute at runtime. In this example, the remote control can be easily extended to support more commands and receivers without changing the existing code.

File: src/Behavioral_Patterns/Command/command.js
// One real world example of the Command design pattern in JavaScript is a remote control for a TV. In this application, the user can use the remote control to turn the TV on/off, change channels, adjust the volume, and perform other actions.

// To implement the Command design pattern in this application, we can create a Command interface that defines a method called execute(). This interface will be implemented by various concrete command classes like PowerCommand, ChannelCommand, VolumeCommand, and MuteCommand.

// Each concrete command class will have its own execute() method that will perform a specific action on the TV. For example, the PowerCommand class will have an execute() method that will turn the TV on or off.

// We can also create a Invoker class that will invoke the execute() method of the concrete command classes based on the user's input. This Invoker class can have a method called setCommand() that will set the concrete command class based on the user's input.

// Here's an example code snippet that demonstrates the Command design pattern in JavaScript for a remote control application:


// Command interface
class Command {
  execute() {}
}

// Concrete command classes
class PowerCommand extends Command {
  constructor(tv) {
    super();
    this.tv = tv;
  }

  execute() {
    this.tv.togglePower();
  }
}

class ChannelCommand extends Command {
  constructor(tv, channel) {
    super();
    this.tv = tv;
    this.channel = channel;
  }

  execute() {
    this.tv.changeChannel(this.channel);
  }
}

class VolumeCommand extends Command {
  constructor(tv, volume) {
    super();
    this.tv = tv;
    this.volume = volume;
  }

  execute() {
    this.tv.adjustVolume(this.volume);
  }
}

class MuteCommand extends Command {
  constructor(tv) {
    super();
    this.tv = tv;
  }

  execute() {
    this.tv.toggleMute();
  }
}

// Invoker class
class RemoteControl {
  setCommand(command) {
    this.command = command;
  }

  pressButton() {
    this.command.execute();
  }
}

// Usage
class TV {
  constructor() {
    this.power = false;
    this.channel = 1;
    this.volume = 50;
    this.mute = false;
  }

  togglePower() {
    this.power = !this.power;
    console.log(`TV is now ${this.power ? 'on' : 'off'}`);
  }

  changeChannel(channel) {
    this.channel = channel;
    console.log(`Changed to channel ${this.channel}`);
  }

  adjustVolume(volume) {
    this.volume = volume;
    console.log(`Volume is now ${this.volume}`);
  }

  toggleMute() {
    this.mute = !this.mute;
    console.log(`Mute is now ${this.mute ? 'on' : 'off'}`);
  }
}

const tv = new TV();
const remote = new RemoteControl();

remote.setCommand(new PowerCommand(tv));
remote.pressButton(); // Output: TV is now on

remote.setCommand(new ChannelCommand(tv, 3));
remote.pressButton(); // Output: Changed to channel 3

remote.setCommand(new VolumeCommand(tv, 70));
remote.pressButton(); // Output: Volume is now 70

remote.setCommand(new MuteCommand(tv));
remote.pressButton(); // Output: Mute is now on


// In this example, we have implemented the Command design pattern to create a remote control application for a TV. The concrete command classes (PowerCommand, ChannelCommand, VolumeCommand, and MuteCommand) perform specific actions on the TV and the Invoker class (RemoteControl) invokes the execute() method of the concrete command classes based on the user's input. This way, we can easily add new actions to the remote control without modifying the existing code.
